### PR TITLE
[FIX] product: allow multiple populate run back to back

### PR DESCRIPTION
**Pull request by rdeodoo**
You can't run the `populate` odoo-bin command more than once if `product.product` is part of the models to be populated. This is because it will always generate a kind of fixed barcode pattern in the form of `BARCODE-PP-{counter}`.

This commit simply performs a quick raw SQL check to ensure  no previous populate run was done, and if it was it simply retrieves the barcode of the last product created through a previous populate command. If that barcode exists, is then used as prefix itself.

Running the populate command 4 times in a row will produce the following results:
```
select id,default_code, barcode from product_product order by id desc;

 id  | default_code |          barcode
-----+--------------+----------------------------
 639 | PP-150       | BARCODE-PP-150-150-150-150
 638 | PP-149       | BARCODE-PP-150-150-150-149
              ...
```

Another option would have been to use a fully randomized string, but avoiding randomness is preferable, especially when considering the case of running the populate command only once.

Note that the `product.product` model is the only problematic one when
running back to back `populate` on all models used by accounting:
- account.move
- account.journal
- account.bank.statement
- account.bank.statement.line
- account.payment
- account.reconcile.model
- account.reconcile.model.line
- res.company
- res.partner
- res.partner.industry
- product.product
- product.category
- ...probably more through `_populate_dependencies`...

task-3821088

### CODE REVIEW

**Review by Feyensv — APPROVED** [review]
Change seems ok (and populate is such a gray area that I don't think it'd annoy anybody).

FYI, there is a rollback parameter to the populate if you only want to test the populate without filling the database with records.  Not useful if what you want is simply to have even more records than the available sizes though :D.

**Review by alialfie — APPROVED** [review]
Hi @rdeodoo , thanks for the PR!
Code looks good to me, my only comment is that we could use the new SQL wrapper here as it's available in 17.0. I don't see any risk of a SQL injection so this is non blocking but it would be nice I think.
@h4818

**Comment by rdeodoo** [comment]
Not sure it's really needed as it's about an hardcoded string in the method but the change is done :+1:

File: addons/product/populate/product.py
Comment (h4818) [inline addons/product/populate/product.py:93]: what about removing this line entirely, and allowing https://github.com/odoo/odoo/blob/17.0/addons/product/populate/product_template.py#L66 to set the barcodes on variants instead?

**Review by h4818 — COMMENTED** [review]
@rdeodoo Thanks for the fix, LGTM!
However, I'm just wondering why we have 2 mechanisms to set barcodes on product variants...
what would happen if we removed this one?

File: addons/product/populate/product.py
Comment (rdeodoo) [reply addons/product/populate/product.py:93]: That code is not related to the issue and would not fix it as it's not called at all since we are fixing the populate of `product.product`, not `product.template` (which is what `account` module is using as `_populate_dependencies`).
The fix has to basically make this to not crash:
```bash
<odoo-bin-command-and-addons> populate -d populate --size=small --models=product.product
<odoo-bin-command-and-addons> populate -d populate --size=small --models=product.product
```

Comment (h4818) [reply addons/product/populate/product.py:93]: My bad, I didn't think of `--models=` and assumed that we'd always get `product.product`s with barcodes anyway.

Anyway, IMO the way that `product.template` sets barcodes on some of it's variants is cleaner and wouldn't lead to longer barcodes on every populate.
From an accounting perspective, we don't really need barcodes, so I leave it up to you guys and the codeowners of `product`

However, I'll just leave my 2 cents and suggest removing `barcode` from the list returned by `_populate_factories` and instead add the barcodes after

```py
def _populate(self, size):
    res = super()._populate(size)

    for product in res:
        product.barcode = "BARCODE-PP-%s" % product.id

    return res
```

Comment (rdeodoo) [reply addons/product/populate/product.py:93]: Writing in a loop, I'd avoid that.
That would make the SQL queries skyrocket, basically 3 extra SQL Queries **per** created product.
While the current solution does not imply any extra sql query.

I checked out of curiosity:

|  | current solution | your solution |
| - | - | - |
| **SELECT Queries** | | |
|  product.product | 0 | 151 |
| product.packaging | 0 | 151 |
| **UPDATE  Queries**|  | |
|  product.product | 1 | 152 |

The populate command being already very slow, I'd avoid that.

**Review by Feyensv — APPROVED** [review]
@robodoo r+ delegate+

**Pull request closed without merging**
